prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt

```diff
@@ -437,60 +418,88 @@ class FHIRRouter(
+                    filterType == ReportStreamFilterType.PROCESSING_MODE_FILTER &&
+                        filters === processingModeFilterDefault
+                    )
+            )
+                "(default filter) "
+            else ""
+            }${failingFilterName ?: "unknown"}"
+            logFilterResults(filterToLog, bundle, report, receiver, filterType)
         }
         return passes
     }
 
     /**
-     * Takes a [bundle] and [filter], evaluates if the bundle passes the filter. If the filter is null,
-     * return [defaultResponse]
+     * Takes a [bundle] and [filter] and optionally a [focusResource], evaluates if the bundle passes the filter, or the
+     * opposite if [reverseFilter] is true. If the filter is null or empty, return [defaultResponse].
      * @return Boolean indicating if the bundle passes the filter or not
-     *          Result will be negated if [reverseFilter] is true
+     *         String to use when logging the filter result
      */
     internal fun evaluateFilterCondition(
         filter: ReportStreamFilter?,
         bundle: Bundle,
         defaultResponse: Boolean,
         reverseFilter: Boolean = false,
-        focusResource: Base = bundle
-    ): Boolean {
+        focusResource: Base = bundle,
+    ): Pair<Boolean, String?> {
         // the filter needs to check all expressions passed in, or if the filter is null or empty it will return the
         // default response
-        val result = if (filter.isNullOrEmpty()) {
-            defaultResponse
-        } else try {
-            filter.all {
-                FhirPathUtils.evaluateCondition(
+        if (filter.isNullOrEmpty()) {
+            return Pair(defaultResponse, "defaultResponse")
+        }
+
+        val failingFilters = mutableListOf<String>()
+        val exceptionFilters = mutableListOf<String>()
+        var result = true
+        filter.forEach { filterElement ->
+            try {
+                val filterElementResult = FhirPathUtils.evaluateCondition(
                     CustomContext(bundle, focusResource, shorthandLookupTable),
                     focusResource,
                     bundle,
-                    it
+                    filterElement
                 )
+                if (!filterElementResult) {
+                    result = false
+                    failingFilters += filterElement
+                }
+            } catch (e: SchemaException) {
+                actionLogger?.warn(
+                    EvaluateFilterConditionErrorMessage(e.message)
+                )
+                exceptionFilters += filterElement
             }
-        } catch (e: SchemaException) {
-            actionLogger?.warn(
-                EvaluateFilterConditionErrorMessage(e.message)
-            )
-            return false
         }
-        return if (reverseFilter) !result else result
+
+        return if (exceptionFilters.isNotEmpty()) {
+            Pair(false, "(exception found) $exceptionFilters")
```

Comment:
You're not actually testing these paths anywhere, are you? Should test that the returned pair is exactly equal to the expected pair for each one of these if/else if branches?

Reply:
Beefed up the tests in an incoming commit.